filecoin/retriever.go

```diff
@@ -461,10 +494,306 @@ func formatCid(cid cid.Cid, short bool) string {
+	retrievalsFinished       int
+}
+
+type activeRetrievalsManager struct {
+	arMap map[cid.Cid]*activeRetrieval
+	lk    sync.RWMutex
+}
+
+func newActiveRetrievalsManager() *activeRetrievalsManager {
+	return &activeRetrievalsManager{arMap: make(map[cid.Cid]*activeRetrieval)}
+}
+
+func (arm *activeRetrievalsManager) New(retrievalCid cid.Cid, queryCandidateCount int, retrievalCandidateCount int) (uuid.UUID, error) {
+	retrievalId, err := uuid.NewRandom()
+	if err != nil {
+		return uuid.UUID{}, err
+	}
+
+	arm.lk.Lock()
+	defer arm.lk.Unlock()
+
+	if arm.findActiveRetrievalFor(retrievalCid) != nil {
+		return uuid.UUID{}, ErrRetrievalAlreadyRunning
+	}
+
+	arm.arMap[retrievalCid] = &activeRetrieval{
+		retrievalId:              retrievalId,
+		retrievalCid:             retrievalCid,
+		rootCid:                  retrievalCid,
+		currentStorageProviderId: "",
+		queryStartTime:           time.Now(),
+		queryCandidateCount:      queryCandidateCount,
+		queriesFinished:          0,
+		retrievalStartTime:       time.Time{},
+		retrievalCandidateCount:  retrievalCandidateCount,
+		retrievalsFinished:       0,
+	}
+
+	log.Debugf("Registered new active retrieval for %s (%d active, %d/%d query candidates)", retrievalCid, len(arm.arMap), 0, queryCandidateCount)
+
+	return retrievalId, nil
+}
+
+func (arm *activeRetrievalsManager) QueryCandidateCount(retrievalCid cid.Cid, candidateCount int) {
+	arm.lk.Lock()
+	defer arm.lk.Unlock()
+
+	ar := arm.findActiveRetrievalFor(retrievalCid)
+	if ar == nil {
+		log.Errorf("Unexpected active retrieval QueryCandidateCount for %s", retrievalCid)
+		return
+	}
+	ar.queryCandidateCount = candidateCount
+	arm.maybeFinish(retrievalCid, ar)
+}
+
+func (arm *activeRetrievalsManager) RetrievalCandidateCount(retrievalCid cid.Cid, candidateCount int) {
+	arm.lk.Lock()
+	defer arm.lk.Unlock()
+
+	ar := arm.findActiveRetrievalFor(retrievalCid)
+	if ar == nil {
+		log.Errorf("Unexpected active retrieval RetrievalCandidateCount for %s", retrievalCid)
+		return
+	}
+	ar.retrievalCandidateCount = candidateCount
+	ar.retrievalStartTime = time.Now()
+	log.Debugf("Updated active retrieval for %s to retrieval phase (%d active, %d/%d query candidates, %d/%d retrieval candidates)", retrievalCid, len(arm.arMap), ar.queriesFinished, ar.queryCandidateCount, ar.retrievalsFinished, ar.retrievalCandidateCount)
+	arm.maybeFinish(retrievalCid, ar)
+}
+
+func (arm *activeRetrievalsManager) StatusFor(retrievalCid cid.Cid, phase rep.Phase) (uuid.UUID, cid.Cid, time.Time, bool) {
+	arm.lk.RLock()
+	defer arm.lk.RUnlock()
+	ar := arm.findActiveRetrievalFor(retrievalCid)
+	if ar == nil {
+		return uuid.UUID{}, cid.Undef, time.Time{}, false
+	}
+	phaseStart := ar.queryStartTime
+	if phase == rep.RetrievalPhase {
+		phaseStart = ar.retrievalStartTime
+	}
+	return ar.retrievalId, ar.retrievalCid, phaseStart, true
+}
+
+func (arm *activeRetrievalsManager) DebugActiveCids(running map[cid.Cid]bool) {
```

Comment:
what is the value of this function? like are we diagnosing something that indicates a larger issue?

Reply:
a debugging piece I'm thinking of removing

Comes back to the problem of making sure that we do clean-up properly - we delete records on failure and success when we have the total number of expected queries+retrievals. But that depends on proper accounting on those numbers **and** making sure we get all of those events. Which it turns out we weren't (see two more failure cases in https://github.com/application-research/filclient/pull/87, plus the timeout case now added in here). I'm slightly concerned that we may not have all of them yet, I can't see obvious cases but I could be missing something, which is why I'm hesitant to entirely remove the debugging stuff.

Reply:
I removed it, I ended up liking the simplicity of less code than the assurance of having it. I have my `awk` script for the log output to check for leaks still and that's fine.